drivers: i2c: Add support for cc32xx I2C

GAnthony opened the pull request:
This was validated on the cc3220sf_launchxl board
using the Zephyr thermometer sample program
adapted to call the i2c driver directly, and fetching
samples from the on-board TMP006 temperature sensor.

A temperature sensor driver for the TMP006 shall follow,
and will be based, on this I2C PR once merged.

- [x] Rework ISR

galak reviewed (changes requested):
In the driverlib import, can we update the commit message to reference the commit b70633b95f46af6138b6e039376af41de6141192
(which has the import source details)

galak commented on boards/arm/cc3220sf_launchxl/board.h:
```diff
@@ -22,4 +22,8 @@
+/* Device Addresses on I2C addresses */
+#define I2C0_TMP_ADDR      (0x41)      /* TMP006 I2C Address */
+#define I2C0_TMP_REGISTER  (0x0001)    /* TMP006 Result Register */
+
```
where are these used?

GAnthony replied:
ah, these were used in a test application, but leaked into this PR.  Will remove.

galak commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+	union dev_config dev_config = (union dev_config)dev_config_raw;
+	u32_t bitrate_id;
+
+	if (!dev_config.bits.is_master_device) {
```
instead of using bits, can we use shifts & masks of the raw value

GAnthony replied:
because ....  these bitfield structures may go away in the future due to the compiler issues discussed in the zephyr mailing list?  OK, best to nip it in the bud.

galak commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+	data->state = I2C_CC32XX_IDLE_MODE;
+
+	/* Take I2C hardware semaphore. */
+	regval = HWREG(0x400F7000);
```
can we cleanup this magic numbers at all?

GAnthony replied:
yup.

galak reviewed (changes requested):
some driver related comments.

mnkp commented on dts/arm/cc3220sf_launchxl.dts:
```diff
@@ -22,3 +23,8 @@
+
+&i2c0 {
+	status = "ok";
+	clock-frequency = <I2C_BITRATE_FAST>;
```
This one is a bit tricky. While I2C module may support the fast speed mode this DTS file is written for an evaluation board. User may choose to attach a standard speed I2C device. I think it's safer to use I2C_BITRATE_STANDARD. If required the I2C module can be later reconfigured to run at a higher speed.

GAnthony replied:
Understand the reasoning behind this request, but would like to keep I2C_BITRATE_FAST here for the following reasons:

* A probable sensors board one would attach to the LaunchXL board is the BOOSTXL-SENSORS booster pack (http://www.ti.com/lit/ug/slau666a/slau666a.pdf); looking at the examples source code in TI Resource Explorer, every sensor example, without exception, sets the rate to 400 kHz when initializing the I2C driver.
* This speed also matches the SimpleLink SDK sensor samples, which is the default rate for the onboard sensors;
* It would be nice for the Zephyr apps/sensor drivers to work with the best rate supported by the hardware, without forcing a reconfiguration call;
* For example, the Zephyr samples/sensor/thermometer sample should (assuming CONFIG_CC32XX_I2C=y) work at the best I2C rate without having to modify the source code specifically for this board.

mnkp replied:
The last two points do not sound very convincing, but the first two do. Maybe we should update board documentation mentioning that the I2C driver is configured in Fast Mode at boot?

GAnthony replied:
Updated board doc.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+		/* Protect prime_transfer from irq: */
+		irq_disable(TI_CC32XX_I2C_40020000_IRQ_0);
+		_i2c_cc32xx_prime_transfer(dev, msgs, addr);
+		irq_enable(TI_CC32XX_I2C_40020000_IRQ_0);
```
It feels strange that you need to disable/enable module IRQ. This is typically handled by enabling masks of specific interrupt sources within the module that you want to handle. Anyway, I'm not familiar with SimpleLink architecture.

GAnthony replied:
Yes, usually that's usually the case.  Will need to take another look at the SimpleLink driverlib. 

Thanks for the review!

GAnthony replied:
Found the masking API, so using that instead.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+	/* Iterate over all the messages */
+	for (int i = 0; i < num_msgs; i++) {
+		/* Acquire the driver mutex */
+		k_sem_take(&data->mutex, K_FOREVER);
```
I believe the mutex should be outside of the loop, otherwise you allow a combined I2C write/read transfer to be interrupted by some unrelated message.

GAnthony replied:
Good point.  The temp sensor driver typically writes a register location first to read, immediately followed by a read op.  Probably a good idea to not interrupt that.

GAnthony commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+	u8_t  *write_buf_idx;  /* Internal inc. writeBuf index */
+	size_t write_cnt;      /* Internal dec. writeCounter */
+	u8_t  *read_buf_idx;   /* Internal inc. readBuf index */
+	size_t read_cnt;       /* Internal dec. readCounter */
```
Agree.

GAnthony commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+		k_sem_take(&data->mutex, K_FOREVER);
+
+		/* Protect prime_transfer from irq: */
+		irq_disable(TI_CC32XX_I2C_40020000_IRQ_0);
```
Seems like an unnecessary complication, given the IRQ number will not change, and comes directly from the dts file.   Other I2C drivers follow this pattern too.  Is there a good reason to make this change?

GAnthony replied:
Made the suggested changes.  But note IRQ_CONNECT will *not* compile with config struct field arguments, so we still end up with a mix of constants and config field arguments in the code.

mnkp replied:
You are right, but that's a known limitation of IRQ_CONNECT. We always put it in a function which is specific to a driver instance.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,431 @@
+
+static const struct i2c_cc32xx_config i2c_cc32xx_config = {
+	.base = TI_CC32XX_I2C_40020000_BASE_ADDRESS,
+	.bitrate = TI_CC32XX_I2C_40020000_CLOCK_FREQUENCY,
```
We should not use the two literals `TI_CC32XX_I2C_40020000_BASE_ADDRESS`, `TI_CC32XX_I2C_40020000_CLOCK_FREQUENCY` directly. Another SoC belonging to the same family and using the same I2C module may place it at another location. This should go through a .fixup file where `_40020000` is removed.

GAnthony replied:
My understanding is fixup files were a temporary work-around - and so have been trying to avoid using them to redefine constants which have already been defined by DTS in the first place.  It's confusing when searching for definitions to see constants being redefined ... but if that's the convention here, OK.

mnkp replied:
There are plans - I'm not sure how advanced - to automatically generate config structures directly from DTS. In this case fixup will not be necessary. Until then however we need this workaround. This would become obvious if you tried to add support for another SoC from the SimpleLink family. Unless all SimpleLink SoCs have only one I2C module and always located at 0x40020000 and no new family member that breaks this convention will ever be manufactured.

Commit:
drivers: i2c: Add support for cc32xx I2C bus

This was validated on the cc3220sf_launchxl board
using the Zephyr thermometer sample program
adapted to call the i2c driver directly, and fetching
samples from the on-board TMP006 temperature sensor.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

GAnthony commented:
@galak, @mnkp, made requested updates.  Thanks for the review!

Note:  I expect a conflict on i2c.h: I2C_SPEED_GET() if PR #1492 is merged first.  In that case, I'll expect to rebase and pull that same definition from my driver.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,417 @@
+struct i2c_cc32xx_config {
+	u32_t base;
+	u32_t bitrate;
+	unsigned int exception_no;
```
Maybe we should change it irq_no? More conventional.

GAnthony replied:
The TI SDK and device documentation consistently uses interrupt numbers in their documentation and device header files.
Zephyr however uses "exception numbers" (which are interrupt numbers shifted down by 16): eg: see ARM technical documentation: http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0203h/Cihbffg3.html .     
Since this is somewhat confusing when writing code which mixes Zephyr API calls that take exception numbers and TI driverlib APIs which use "interrupt numbers", the variable name here is meant to clarify the true meaning of the value.

GAnthony replied:
Hmm, seems it's the other way around: TI uses exception numbers. OK, will change to irq_no.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,417 @@
+
+static struct i2c_cc32xx_data i2c_cc32xx_data;
+
+DEVICE_AND_API_INIT(i2c_cc32xx, TI_CC32XX_I2C_40020000_LABEL, &i2c_cc32xx_init,
```
TI_CC32XX_I2C_40020000_LABEL should also go through fixup.

GAnthony replied:
Yup, missed that one.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,417 @@
+	struct k_sem mutex;
+	struct k_sem transfer_complete;
+
+	volatile enum i2c_cc32xx_state state;
```
It seems like the state is not really used in the driver code, there are no state machine like dependencies. We only differentiate between normal operation and an error condition. We could replace this enum with a single flag.

GAnthony replied:
Actually, the isr has a state machine which determines the next path depending on how the I2C transfer was primed: for read or write.  The distinction of read vs write could have been determined from the read or write index variables, but we coalesced those into one.   It seems the state variable is still needed here.

mnkp replied:
You are right, I completely overlooked the `switch (data->state)` statement in the ISR. That said, I think we should have a closer look at the ISR. Your current implementation may work well but has some weaknesses. 

The typical way to implement an ISR would be to react to specific status flags ISR was invoked with. I.e. the conditional statement should be built around value of I2CMMIS "I2C Master Masked Interrupt Status" register (I gave only a brief look into the datasheet, but I think I got it right). Often more than one flag can be activated simultaneously and checking the flags directly is the most reliable way to determine the reason ISR was invoked. If you attempt to infer the cause of ISR from other sources, e.g. the data->state variable there is a chance you get it wrong. What if some other interrupt mask was enabled which was not taken into account? Some of the other interrupts may happen only under specific conditions. Such errors will not easily be caught under basic testing conditions.

Using data->state variable to determine ISR cause makes it also difficult to extend the driver in the future. A direct dependency on interrupt status is much easier to work with when new features are added.

GAnthony replied:
That seems reasonable.  Will need to look at modifying the logic in the next iteration.

GAnthony replied:
@mnkp, @galak, finally got back to this task, and made recommended changes to the ISR.

The ISR now is (mainly) driven by the interrupt status bits.  
Since in this driver, FIFOs are not directly enabled and used, I found no interrupt registers telling whether the data transfer was primed as a read or a write, so I kept those state variables only for that purpose.  Everything else is driven by the ISR flags.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,417 @@
+			 */
+			MAP_I2CMasterControl(base,
+					  I2C_MASTER_CMD_BURST_SEND_ERROR_STOP);
+			k_sem_give(&data->transfer_complete);
```
It looks like this line will be executed regardless of the if condition. We can put it after the if statement.

GAnthony replied:
Yes, done.

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,417 @@
+	for (int i = 0; i < num_msgs; i++) {
+
+		/* Protect prime_transfer from irq: */
+		MAP_I2CMasterIntDisable(base);
```
If we are here it means that the last transfer has finished and the interrupts should have already been disabled. If they are not the ISR routine will attempt to handle them even if the driver itself isn't expecting it. We should not need this line. This is just a suggestion to double check the concept. I may be wrong.

GAnthony replied:
Yes, ideally, shouldn't be needed.  Removed the protection.

dbkinder commented on boards/arm/cc3220sf_launchxl/doc/cc3220sf_launchxl.rst:
```diff
@@ -56,6 +56,12 @@ driver support.
+| I2C       | on-chip    | i2c                   |
++-----------+------------+-----------------------+
+
+Note: For consistency with TI SimpleLink SDK and BoosterPack examples,
```
If you want this note to stand out more, you could use this:

     .. note::
 
        For consistency with TI SimpleLink SDK and BoosterPack examples,
        the I2C driver defaults to I2C_BITRATE_FAST mode (400 kHz) bus
        speed on bootup.

GAnthony replied:
OK, added the .. note:: directive.

dbkinder reviewed (dismissed):
note directive suggestion, but OK as is.

galak commented:
@mnkp any other comments, all changes good?

mnkp commented:
I think it would be nice to rework the ISR routine. Otherwise it looks good.

GAnthony commented:
Agree ... will try to find time to rework the ISR this week.

galak commented:
Wondering if I can make a markdown todo list with a comment:

- [ ] Rework ISR

galak commented:
ping, hoping to close this out soon

mnkp commented on drivers/i2c/i2c_cc32xx.c:
```diff
@@ -0,0 +1,395 @@
+		/*
+		 * No more data needs to be sent, so follow up with
+		 * a STOP bit.
+		 * Again, there is no equivalent macro (0x00000004) so
```
I'm not sure this comment is helpful, maybe we should remove it. There are two more places where it's used. The lines talking about "no equivalent macro".

GAnthony replied:
OK, will remove.

mnkp reviewed (approved):
LGTM, with some minor remark regarding comments. May just as well be ignored.

GAnthony commented:
@galak, @mnkp, comments about macros removed.   Ready to merge.
Thanks!
